### PR TITLE
cdc, externalconn: enable external connections for all cdc sinks

### PR DESCRIPTION
**Pull request by HonoreDB**
This PR expands the kafka external connection sink to a generalized
external connection sink for all changefeed sink types. Because these
overlap with backup sink types, this necessitates changing the external
connection registration logic slightly: different consumers of the same scheme need to use the same parsing logic, but can have different validation logic.

Informs https://github.com/cockroachdb/cockroach/issues/88928

### CODE REVIEW

**Comment by adityamaru** [comment]
Thanks for pushing this out, is the motivation for the duplication of implementations for the shared external storage schemes i.e. azure, gs, s3, nodelocal, userfile that changefeeds and external storage differ in their validation? Currently for all external storage implementations we write, list and read a sentinel file to ensure that the URI is reachable. Is this not sufficient for changefeeds since I assume they use the same `ExternalStorage` constructor as backup etc.

I'm not sure how I feel about the 1:N scheme to parse function mapping, my hope was that we wouldn't have to re-register ES providers for CDC.

**Comment by miretskiy** [comment]
> I'm not sure how I feel about the 1:N scheme to parse function mapping,

My hope the same.

**Comment by miretskiy** [comment]
@HonoreDB @adityamaru : If I recall correctly, the think was that on cdc side, we will simply get external connection record, unwrap that record (to get URI) and just recursively call "getSink".  I don't understand why we would need 1:N stuff.

**Comment by HonoreDB** [comment]
We should probably have a call to figure this out, yeah. My goal with the 1:N thing was to avoid tight coupling between different users of the same external connection URI scheme, to the extent that's possible. Scenarios I'm thinking about:

1. We register "http" in CDC. Later, Bulk IO decides they also want to support backup over http as a non-CCL feature. If we only allow one registration, they'll probably need to delete CDC's registration and register their own, since they can't count on the CDC registration happening (e.g. if they're in a binary created by `make build oss`.) Similarly, CDC now has functionality that effectively depends on that registration code in an unrelated file. If we made the method RegisterIfNotExists(scheme), then one of the two registrants has the ability to break the other one if it changes its parsing logic.
2. Changefeeds have validations that don't apply to other uses of the same scheme. Right now we error out if you try to create a gcpubsub sink without specifying a region parameter, but that's for changefeed-specific reasons that other users of that external connection shouldn't be bound by. Conversely, some other user of gcpubsub connections might need you to specify a consumer-managed encryption key at create time, which changefeeds don't require. So if you create an external connection that has neither region nor cmek, should the create succeed even though nobody can use it?

So my proposal here is to let two registrations for the same scheme coexist independently, with the only interaction being that we ensure there's no URI that they both consider valid but would serialize as a protobuf differently (and if that ever became a possibility, we'd just start storing up to N protobufs per external connection).

I think another approach to consider could be to remove the idea of registering custom validate/parse functions entirely, and have a finite set of usage-agnostic parsing options defined in external connection code, so all registrants need to do is at most make sure they're picking the same enum value for the same scheme.

**Comment by adityamaru** [comment]
> and have a finite set of usage-agnostic parsing options

For External Storage, and KMS we had a lengthy discussion about how we would like to front load a lot of the validation. So at this point we go through the motions of actually writing, reading, listing, encrypting, decrypting a sentinel file instead of just parse time validation.

**Comment by adityamaru** [comment]
This might be suboptimal but if two implementations of a scheme differ enough in their validation maybe they should be represented by two different scheme prefixes? Given this is a new feature we could also consider introducing new schemes such as `cdc-http` or something to decouple the implementations.

**Comment by HonoreDB** [comment]
Sorry I let this sit a bit after our last round of design discussions. Updated, this is ready for another look.

**Comment by adityamaru** [comment]
Sorry for the delay, going to give this a look today @HonoreDB but might be worth getting someone on CDC for the changefeed parts!

File: pkg/cloud/externalconn/impl_registry.go [hunk 1]
```diff
@@ -16,23 +16,86 @@ import (
...
+	connectionParserFactory
+	validations map[string]ValidationFn
+}
+
 // parseAndValidateFns maps a URI scheme to a constructor of instances of that external
 // connection.
-var parseAndValidateFns = map[string]connectionParserFactory{}
+var parseAndValidateFns = map[string]schemeRegistration{}
+
+// FactoryType identifies the expected parsing logic used to create an external connection.
```
Comment (adityamaru) [inline pkg/cloud/externalconn/impl_registry.go:35]: nit: I prefer having all consts and type declarations at the top of the file, but I leave it up to you 🙂

File: pkg/cloud/externalconn/impl_registry.go [hunk 2]
```diff
@@ -44,13 +107,35 @@ func ExternalConnectionFromURI(
...
 	}
 
-	return parseFn(ctx, execCfg, user, externalConnectionURI)
+	return parseAndValidateFn.parseAndValidateURI(ctx, execCfg, user, externalConnectionURI, defaultValidation)
+}
+
+// ValidationFn is the type taken by Register*Validation. Validation functions can ping the external connection.
+type ValidationFn func(ctx context.Context, execCfg interface{}, user username.SQLUsername, uri string) error
+
+func (s *schemeRegistration) parseAndValidateURI(
```
Comment (adityamaru) [inline pkg/cloud/externalconn/impl_registry.go:122]: nit: can we move this method under the struct decalaration

File: pkg/ccl/changefeedccl/helpers_test.go [hunk 3]
```diff
@@ -860,14 +860,33 @@ func cdcTestNamedWithSystem(
 	t.Run(testLabel, func(t *testing.T) {
 		testServer, cleanupServer := makeServerWithOptions(t, options)
 		feedFactory, cleanupSink := makeFeedFactoryWithOptions(t, sinkType, testServer.Server, testServer.DB, options)
+		feedFactory = maybeUseExternalConnection(feedFactory, testServer.DB, sinkType, options, t)
 		defer cleanupServer()
 		defer cleanupSink()
 		defer cleanupCloudStorage()
-
+		t.Log()
```
Comment (adityamaru) [inline pkg/ccl/changefeedccl/helpers_test.go:867]: nit: leftover debug

File: pkg/cloud/externalconn/impl_registry.go [hunk 4]
```diff
@@ -16,23 +16,86 @@ import (
...
+					URI: uri.String(),
+				},
+			},
+		}
+		return NewExternalConnection(connDetails), nil
 	}
-	parseAndValidateFns[providerScheme] = parseAndValidateFn
 }
 
 // ExternalConnectionFromURI returns a ExternalConnection for the given URI.
```
Comment (adityamaru) [inline pkg/cloud/externalconn/impl_registry.go:101]: Can I ask you to flesh out this comment to reflect that it creates an External Connection and runs the registered validation which could include pinging the resource.

**Review by adityamaru — CHANGES_REQUESTED** [review]
I've reviewed all the EC parts and they look nice! Only nits and one question from me, but I'd get someone from CDC to review the first commit.

In your second commit message you mention "this will just do the default validation" but say we run a `CREATE EXTERNAL CONNECTION foo AS 'userfile:///foo'`, my understanding is that this will run all the validations i.e. default registered by `pkg/cloud/filetable` and the `changefeed` validation registered in this diff? Or am I missing something. The loop in `parseAndValidateURI` over `validations` doesn't seem to restrict what validation we run.

**Review by HonoreDB — COMMENTED** [review]
Thanks for the reviews! Addressed your comments. The loop in question is over the validations passed as arguments, not the validations on the scheme, and currently that's hardcoded to only be the default validation. I changed the variable name to make that clearer.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86061#-:-NAtTzYk4XT904ih4WN1:bjjkq7a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @HonoreDB, and @shermanCRL)

___
*[`pkg/cloud/externalconn/impl_registry.go` line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86061#-N9Itox1hI9nx8XxoZSx-r5-35:-NAtRpWkAaYxrtSoxyXm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec7a4e31e360c6c7bb615c0c74bf7220fbf6b4f4/pkg/cloud/externalconn/impl_registry.go#L35)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: I prefer having all consts and type declarations at the top of the file, but I leave it up to you 🙂 
</blockquote></details>

Done.
___
*[`pkg/cloud/externalconn/impl_registry.go` line 66 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86061#-N9Itox1hI9nx8XxoZSx-r5-66:-NAtU24BEuZz9Px3yB5k:bmv44jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec7a4e31e360c6c7bb615c0c74bf7220fbf6b4f4/pkg/cloud/externalconn/impl_registry.go#L66)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Should we put all the validation identifiers as consts in this file, the `changefed` one too.
</blockquote></details>

I'm sure we're going to change this around soon but right now only the default identifier does anything so I'd rather have that be the only constant.
___
*[`pkg/cloud/externalconn/impl_registry.go` line 101 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86061#-N9Itox1hI9nx8XxoZSx-r5-101:-NAtUC3E1ig8ep2RF8xO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec7a4e31e360c6c7bb615c0c74bf7220fbf6b4f4/pkg/cloud/externalconn/impl_registry.go#L101)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Can I ask you to flesh out this comment to reflect that it creates an External Connection and runs the registered validation which could include pinging the resource.
</blockquote></details>

Done.

File: pkg/cloud/externalconn/connectionpb/connection.go [hunk 5]
```diff
@@ -23,7 +23,7 @@ func (d *ConnectionDetails) Type() ConnectionType {
 	case ConnectionProvider_kafka:
 		return TypeStorage
 	default:
-		panic(errors.AssertionFailedf("ConnectionDetails.Type called on a details with an unknown type: %s", d.Provider.String()))
```
[the remark this answers is not in the document]
Comment (HonoreDB) [reply pkg/cloud/externalconn/connectionpb/connection.go:26]: Oops, missed this comment, I'll fix this now.

**Review by jayshrivastava — APPROVED** [review]
> Reviewed 8 of 21 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86061)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @HonoreDB, @jayshrivastava, and @shermanCRL)

**Review by adityamaru — APPROVED** [review]
🚀

**Comment by HonoreDB** [comment]
Yup, sorry, still trying to figure out some test failures.

**Comment by HonoreDB** [comment]
bors retry

**Pull request merged**